### PR TITLE
[SPARK-21584][SQL][SparkR] Update R method for summary to call new implementation

### PR DESCRIPTION
**Pull request by aray**
## What changes were proposed in this pull request?

SPARK-21100 introduced a new `summary` method to the Scala/Java Dataset API that included  expanded statistics (vs `describe`) and control over which statistics to compute. Currently in the R API `summary` acts as an alias for `describe`. This patch updates the R API to call the new `summary` method in the JVM that includes additional statistics and ability to select which to compute.

This does not break the current interface as the present `summary` method does not take additional arguments like `describe` and the output was never meant to be used programmatically. 

## How was this patch tested?

Modified and additional unit tests.

### CODE REVIEW

File: R/pkg/R/DataFrame.R [hunk 1]
```diff
@@ -2973,15 +2974,51 @@ setMethod("describe",
...
+#'
+#' Computes specified statistics for numeric and string columns.
+#'
+#' Available statistics are:
+#'
+#' - count
+#' - mean
+#' - stddev
+#' - min
+#' - max
```
Comment (felixcheung) [inline R/pkg/R/DataFrame.R]: these bullets and whitespaces get collapsed by roxgyen2 - try itemize/item https://cran.r-project.org/web/packages/roxygen2/vignettes/rd.html

File: R/pkg/R/DataFrame.R [hunk 2]
```diff
@@ -2973,15 +2974,51 @@ setMethod("describe",
             dataFrame(sdf)
           })
 
+#' summary
+#'
+#' Computes specified statistics for numeric and string columns.
```
Comment (felixcheung) [inline R/pkg/R/DataFrame.R]: does this apply to "string columns"?

File: R/pkg/tests/fulltests/test_sparkSQL.R [hunk 3]
```diff
@@ -2500,8 +2500,15 @@ test_that("describe() and summarize() on a DataFrame", {
   expect_equal(collect(stats)[5, "age"], "30")
 
   stats2 <- summary(df)
-  expect_equal(collect(stats2)[4, "summary"], "min")
-  expect_equal(collect(stats2)[5, "age"], "30")
+  expect_equal(collect(stats2)[5, "summary"], "25%")
+  expect_equal(collect(stats2)[5, "age"], "30.0")
```
Comment (felixcheung) [inline R/pkg/tests/fulltests/test_sparkSQL.R]: does this mean this change the output of `summary(df)` call?

File: R/pkg/R/DataFrame.R [hunk 4]
```diff
@@ -2973,15 +2974,51 @@ setMethod("describe",
...
+#'
+#' Available statistics are:
+#'
+#' - count
+#' - mean
+#' - stddev
+#' - min
+#' - max
+#' - arbitrary approximate percentiles specified as a percentage (eg, 75%)
```
Comment (felixcheung) [inline R/pkg/R/DataFrame.R]: I'd clarify that `75%` should be a string, eg. `"75%"`

File: R/pkg/R/DataFrame.R [hunk 5]
```diff
@@ -2973,15 +2974,51 @@ setMethod("describe",
...
+#'
+#' If no statistics are given, this function computes count, mean, stddev, min,
+#' approximate quartiles (percentiles at 25%, 50%, and 75%), and max.
```
Comment (felixcheung) [inline R/pkg/R/DataFrame.R]: also, don't use empty line - like #' - the 2nd one becomes the "details" section in the doc as formatted by roxygen2

File: R/pkg/tests/fulltests/test_sparkSQL.R (hunk 3 above)
Comment (aray) [inline R/pkg/tests/fulltests/test_sparkSQL.R]: Yes

File: R/pkg/R/DataFrame.R (hunk 2 above)
Comment (aray) [inline R/pkg/R/DataFrame.R]: This is unchanged from before. The stats are only computed for `NumericType` and `StringType` columns. Of course the only ones that are non null for strings are count, min, and max.

**Comment by HyukjinKwon** [comment]
@aray, it looks the tests with AppVeyor failed due to time limit, 1.5 hours. Would you mind closing and reopening this one to retrigger the test?

@felixcheung, It sounds now we are sometimes reaching the limit again... I requested to increase this to AppVeyor from 1 to 1.5 hours before but sounds we should figure out another way ... Now it looks now taking roughly 1.1 ~ 1.5 hours.

**Comment by felixcheung** [comment]
@HyukjinKwon I'm not sure how - in AppVeyor we are building everything from scratch... it does take time

**Review by felixcheung — COMMENTED** [review]
Let's track behavior/output changes in release note/migration guide.

opened https://issues.apache.org/jira/browse/SPARK-21616

@aray when the output changes to summary, is it "additive"? ie. at the end?

**Comment by aray** [comment]
No the changes to `summary` are not additive, it inserts 25%, 50%, and 75% percentiles before max (the last row). People that want the previous behavior can use `describe`. Or if they are trying to programmatically access these fields they should really be explicitly specifying an aggregation. If you recall we discussed using the summary name in the original PR https://github.com/apache/spark/pull/18307#issuecomment-312347948

**Comment by felixcheung** [comment]
I see. I recall the method name discussion; though changing API and/or output format is something we generally want to avoid. Something like this has been called out in past releases as we shouldn't do in the future.

**Comment by felixcheung** [comment]
Is it too late to change the Scala side output format? I suspect it doesn't matter too much on Scala/Python which order they are in and preserving the existing order in R could be helpful.

**Comment by aray** [comment]
@rxin Any thoughts on whether it's ok to change the output of `summary` in R in a non "additive" way?

**Comment by rxin** [comment]
I suspect it is ok for R ...

**Comment by aray** [comment]
I'm pushing for it to stay as is because it's the more logical layout of the data: min=0%, 25%, 50%, 75%, max=100%. It's also more consistent with summary of native R dataframes (and for Python the Pandas describe method).

Anyone who is accessing these fields blindly by index should know they are taking a risk. Furthermore we already cast everything to strings for the output of summary so it should be obvious that it's not meant for reuse.

@felixcheung If you still feel strongly a compromise might be to print a warning when `summary` is called from R with no additional arguments.

File: R/pkg/R/DataFrame.R [hunk 6]
```diff
@@ -2953,6 +2953,7 @@ setMethod("saveAsTable",
 #' describe(df, "col1")
 #' describe(df, "col1", "col2")
 #' }
+#' @seealso Ues \code{\link{summary}} for expanded statistics and control over which statistics to compute.
```
Comment (felixcheung) [inline R/pkg/R/DataFrame.R]: also, I think `\link{summary}}` is sufficient, no need for `\code`

File: R/pkg/R/DataFrame.R [hunk 7]
```diff
@@ -2973,15 +2974,48 @@ setMethod("describe",
...
+#' sparkR.session()
+#' path <- "path/to/file.json"
+#' df <- read.json(path)
+#' summary(df)
+#' summary(df, "min", "25%", "75%", "max")
+#' summary(select(df, "age", "height"))
+#' }
 #' @note summary(SparkDataFrame) since 1.5.0
+#' @note The statistics provided by \code{summary} were change in 2.3.0 use \code{\link{describe}} for previous defaults.
+#' @seealso \code{\link{describe}}
```
Comment (felixcheung) [inline R/pkg/R/DataFrame.R]: ditto here and the previous line with `\code`

File: R/pkg/R/DataFrame.R [hunk 8]
```diff
@@ -2973,15 +2974,48 @@ setMethod("describe",
...
+#' \item stddev
+#' \item min
+#' \item max
+#' \item arbitrary approximate percentiles specified as a percentage (eg, "75%")
+#' }
+#' If no statistics are given, this function computes count, mean, stddev, min,
+#' approximate quartiles (percentiles at 25%, 50%, and 75%), and max.
+#' This function is meant for exploratory data analysis, as we make no guarantee about the
+#' backward compatibility of the schema of the resulting Dataset. If you want to
+#' programmatically compute summary statistics, use the `agg` function instead.
```
Comment (felixcheung) [inline R/pkg/R/DataFrame.R]: oh and don't use backtick - it doesn't get process by roxygen2
use `\code{agg}` instead

**Comment by aray** [comment]
closing and reopening to trigger AppVeyor test that timed out

**Review by felixcheung — APPROVED** [review]
LGTM

**Comment by felixcheung** [comment]
merged to master

**Pull request closed without merging**